src/number.rs

```diff
@@ -0,0 +1,165 @@
+//! TODO: Add documentation when approved
+
+pub use lexical::format;
+
+use crate::extra::ParserExtra;
+use crate::input::{InputRef, SliceInput};
+use crate::private::{Check, Emit, Mode, PResult, ParserSealed};
+use crate::EmptyPhantom;
+
+use lexical::parse_partial;
+use lexical::FromLexical;
+
+/// TODO: Add documentation when approved
+#[derive(Clone, Copy)]
+pub struct Number<const F: u128, I, E, O> {
+    #[allow(dead_code)]
+    phantom: EmptyPhantom<(I, E, O)>,
+}
+
+/// TODO: Add documentation when approved
+pub const fn number<const F: u128, I, E, O>() -> Number<F, I, E, O> {
+    Number::<F, I, E, O> {
+        phantom: EmptyPhantom::new(),
+    }
+}
+
+impl<'a, const F: u128, I, E, O> ParserSealed<'a, I, O, E> for Number<F, I, E, O>
+where
+    O: FromLexical,
+    I: SliceInput<'a, Offset = usize>,
+    <I as SliceInput<'a>>::Slice: AsRef<[u8]>,
+    E: ParserExtra<'a, I>,
+{
+    #[inline]
+    fn go<M: Mode>(&self, inp: &mut InputRef<'a, '_, I, E>) -> PResult<M, O> {
+        let before = inp.offset();
+        match parse_partial(inp.slice_trailing_inner().as_ref()) {
+            Ok((out, skip)) => {
+                inp.skip_bytes(skip);
+                Ok(M::bind(|| out))
+            }
+            Err(_err) => {
+                // TODO: Improve error
+                inp.add_alt(inp.offset().offset, None, None, inp.span_since(before));
+                Err(())
+            }
+        }
+    }
+
+    go_extra!(O);
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::{extra, Parser};
+    use lexical::format::RUST_LITERAL;
+
+    // These have been shamelessly yanked from the rust test-float-parse suite.
+    // More specifically:
+    //
+    // https://github.com/rust-lang/rust/tree/64185f205dcbd8db255ad6674e43c63423f2369a/src/etc/test-float-parse
+    mod rust {
+        use super::*;
+
+        const FLOAT: Number<RUST_LITERAL, &str, extra::Default, f64> = number();
```

Comment:
It's a bit of a shame that we end up needing to specify the type parameters like `number::<RUST_LITERAL, _, _, _>()` each time due to the const generic, but by the looks of things there might not be a way around this.

Reply:
I agree, but I have no clue how we could even try to work around this, because `parse_partial` uses the `RUST_LITERAL` generically.

Unrelated, but It may also be worth it to use `const_assert` to make sure that the literal provided is [valid](https://docs.rs/lexical/latest/lexical/fn.format_is_valid.html).